### ISSUE TITLE
Wil CoreWCF be able to encrypt across the wire

### ISSUE THREAD

**Issue by dgxhubbard** [CoreWCF/CoreWCF#81]
I don't know if its to early to ask this question, but will CoreWCF support encryption across the wire over tcp or will it support https using kestrel. I don't really care about what is be it JSON or so on. I just want to know will be able to add encryption so that things are not sent plain text over the wire.

**Comment by mconnew** [comment]
It already does. For HTTP, CoreWCF is just middleware on top of aspnetcore. So set up aspnetcore to use HTTPS and CoreWCF just uses it. If you are using NetTcp, it supports Windows credentials (which uses NegotiateStream and is encrypted) or Certificate credentials (which uses SslStream so is encrypted).

**Comment by dgxhubbard** [comment]
Great thanks.

**Comment by dgxhubbard** [comment]
After some thought I am confused on both http and tcp. We are developing windows apps and have not used asp.net core. For http, can I use a book on asp.net core to get an example? Can I do the same for NetTcp?

**Comment by mconnew** [comment]
For HTTP, all you need to do is create a regular asp.net core app without doing anything special when creating the `IWebHostBuilder` other than you need to add `.UseUrls("http://localhost:8080")` or whatever your port number you need is. I plan to try and work out how to get rid of that, but for now it's needed.
If you want to use NetTcp, you can only use the Kestrel host as CoreWCF uses that for NetTcp and asp.net core only supports a single host implementation. You will need to add the NetTcp handler to your `IWebHostBuilder` similar to [this sample](https://github.com/CoreWCF/CoreWCF/blob/master/src/Samples/NetCoreServer/Program.cs#L20). That sample uses port 8808 but the default port for NetTcp is usually 808. This was in case someone had WCF Tcp port sharing enabled on their machine (or an actual WCF service) using port 808 which would cause the sample to break.  
Then in your Startup class you [need to add CoreWCF](https://github.com/CoreWCF/CoreWCF/blob/master/src/Samples/NetCoreServer/Program.cs#L20) to your application services in your `ConfigureServices` method. Then you add your service and listening endpoints (and bindings) to you `Configure` method similar to https://github.com/CoreWCF/CoreWCF/blob/master/src/Samples/NetCoreServer/Startup.cs#L18.  
As much as possible has the same api surface as WCF on .NET Framework. For example, the NetTcpBinding has the same api surface for configuring it as on .NET Framework. The only difference is namespace.  
CoreWCF is build on top of ASP.NET Core 2.1 and likely won't work on top of ASP.NET Core 3.x as they broke a bunch of api's. ASP.NET Core 2.1 will run on .NET Framework which is why CoreWCF hasn't moved to ASP.NET Core 3.x yet. You can still run everything on .NET Core 3.0, but you must use the older ASP.NET Core 2.1.  
I can't point you to any specific books or blog posts I'm afraid. I learned about how everything works by reading the asp.net core source code. Any documentation I could find to point you to seemed like it presumed you already knew all the other stuff about asp.net core first. You could start with the sample app I pointed you to and you could go from there. I think we probably need some cookbook type documentation for various scenarios.

**Comment by dgxhubbard** [comment]
"CoreWCF is built on top of ASP.NET Core 2.1 and won't work on top of ASP.NET Core 3.x."
Will that change later in the year? We just switched everything over to .net core 3.1 with WPF.

**Comment by dgxhubbard** [comment]
I can contribute work to the conversion, would need help in what to do. Could it be a separate repo?

**Comment by mconnew** [comment]
It will work on .NET Core 3.1. It's the **ASP.NET Core** version that is older.

**Comment by mconnew** [comment]
And at some point there will be support for later versions ASP.NET Core. But after a 1.0 release is working on .NET Framework and .NET Core to help migration scenarios.

**Comment by dgxhubbard** [comment]
Sorry I am confused. I thought asp.net core 3.1 comes with net core 3.1. I looked at my dependencies and most asp.net core libraries are at 2.1.X so am I ok to run?

**Comment by mconnew** [comment]
ASP.NET Core 3.1 does come with .NET Core 3.1, but you don't have to use it. It's enabled by setting your SDK on the project element to `Microsoft.NET.Sdk.Web`. For example:
```xml
<Project Sdk="Microsoft.NET.Sdk.Web">
```
This automatically pulls in all the ASP.NET Core assemblies which come with .NET Core. This is what won't work with CoreWCF. If instead you specify the SDK to be `Microsoft.NET.Sdk`, then no ASP.NET Core libraries which ship with .NET Core get pulled in. So your project line would look like this:
```xml
<Project Sdk="Microsoft.NET.Sdk">
```
Then when you add a package reference to ASP.NET Core 2.1, it's being used just like a third party library like it was on earlier versions of .NET Core. The evolution of the ASP.NET Core packaging story can make it a bit confusing. 
Basically, just use the .NET Core only sdk and add the ASP.NET Core 2.1 packages as package references and it will all work.

**Comment by dgxhubbard** [comment]
sound great thanks! appreciate all the work you are doing.

**Comment by KamranShahid** [comment]
> It will work on .NET Core 3.1. It's the **ASP.NET Core** version that is older.

Sample doesn't working. trying following in my .net core 3.1 
https://github.com/CoreWCF/CoreWCF/blob/master/src/Samples/NetCoreServer/Startup.cs
